src/vm/costs/mod.rs

```diff
@@ -220,57 +270,70 @@ fn parse_cost(cost_function: ClarityCostFunction, eval_result: InterpreterResult
-        Err(e) => Err(CostErrors::CostComputationFailed(format!("Error evaluating result of cost function {}: {}", cost_function.get_name(), e))),
+        }
+        Ok(Some(_)) => Err(CostErrors::CostComputationFailed(
+            "Clarity cost function returned something other than a Cost tuple".to_string(),
+        )),
+        Ok(None) => Err(CostErrors::CostComputationFailed(
+            "Clarity cost function returned nothing".to_string(),
+        )),
+        Err(e) => Err(CostErrors::CostComputationFailed(format!(
+            "Error evaluating result of cost function {}: {}",
+            cost_function.get_name(),
+            e
+        ))),
     }
 }
 
 fn compute_cost(
     cost_tracker: &mut LimitedCostTracker,
     cost_function: ClarityCostFunction,
-    input_size: u64) -> Result<ExecutionCost> {
-
+    input_size: u64,
+) -> Result<ExecutionCost> {
     let mut null_store = NullBackingStore::new();
     let conn = null_store.as_clarity_db();
     let mut global_context = GlobalContext::new(conn, LimitedCostTracker::new_free());
 
-    let cost_function_reference = cost_tracker.cost_function_references
+    let cost_function_reference = cost_tracker
+        .cost_function_references
         .get(&cost_function)
-        .ok_or(CostErrors::CostComputationFailed("CostFunction not defined".to_string()))?.clone();
-
-    let cost_contract = cost_tracker.cost_contracts
-        .get(&cost_function_reference.contract_id)
-        .ok_or(CostErrors::CostComputationFailed("Cost Contract not cached".to_string()))?;
+        .ok_or(CostErrors::CostComputationFailed(
+            "CostFunction not defined".to_string(),
+        ))?
+        .clone();
 
-    let mut contract_context = ContractContext::new(cost_function_reference.contract_id.clone());
+    let cost_contract = cost_tracker
+        .cost_contracts
+        .get_mut(&cost_function_reference.contract_id)
+        .ok_or(CostErrors::CostComputationFailed(
+            "Cost Contract not cached".to_string(),
+        ))?;
 
-    let mut expressions = cost_contract.expressions.clone();
-    let mut program = vec![SymbolicExpression::atom(cost_function_reference.function_name[..].into())];
-    program.push(SymbolicExpression::atom_value(Value::UInt(input_size.into())));
+    let program = vec![
+        SymbolicExpression::atom(cost_function_reference.function_name[..].into()),
+        SymbolicExpression::atom_value(Value::UInt(input_size.into())),
+    ];
 
-    let function_invocation = SymbolicExpression::list(program.into_boxed_slice());
-    expressions.append(&mut vec![function_invocation]);
+    let function_invocation = [SymbolicExpression::list(program.into_boxed_slice())];
 
-    let eval_result = global_context.execute(|g| {
-        eval_all(&expressions, &mut contract_context, g)
-    });
+    let eval_result = eval_all(&function_invocation, cost_contract, &mut global_context);
```

Comment:
Something occurred to me when I was reading this line:  a cost function can assign different costs for a function to different principals, based on who the `sender` and `caller` are.  Is this something we want to support, or something we should prohibit?  The case for prohibiting is that it could lead to a world where miners vote themselves (or their paying clients) discounts for all functions, and get to run smart contracts for as cheap as possible.

Reply:
I'd say we definitely want to prevent that -- I'd recommend adding this comment to #2041, as this is another kind of constraint we want to apply to these functions.